lib/src/model/relation/relation_ctrl.dart

```diff
@@ -0,0 +1,87 @@
+import 'dart:async';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:lichess_mobile/src/model/auth/auth_socket.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+import 'package:lichess_mobile/src/model/common/socket.dart';
+
+part 'relation_ctrl.freezed.dart';
+part 'relation_ctrl.g.dart';
+
+@riverpod
+class RelationCtrl extends _$RelationCtrl {
+  StreamSubscription<SocketEvent>? _socketSubscription;
+
+  @override
+  Future<RelationCtrlState> build() {
+    final socket = ref.watch(authSocketProvider);
+    final (stream, _) = socket.connect(Uri(path: '/lobby/socket/v5'));
+
+    final state = stream.firstWhere((e) => e.topic == 'following_onlines').then(
+      (event) {
+        _socketSubscription = stream.listen(_handleSocketTopic);
+        return RelationCtrlState(followingOnlines: event.data as List<dynamic>);
+      },
+    );
+
+    ref.onDispose(() {
+      _socketSubscription?.cancel();
+    });
+
+    return state;
+  }
+
+  void getFollowingOnlines() {
+    _socket.send('following_onlines', null);
+  }
+
+  void _handleSocketTopic(SocketEvent event) {
+    switch (event.topic) {
+      case 'following_onlines':
+        final data = event.data as List<dynamic>;
+        if (state is AsyncData) {
```

Comment:
instead of checking the type and cast, it's best to make a single check at the top of the `_handleSocketTopic`:
```dart
void _handleSocketTopic(SocketEvent event) {
  if (!state.hasValue) return;
 
  ...
  // then you can write:
  state = AsyncValue.data(state.requireValue.copyWith(followingOnlines: newData));
  ...
}
```